#157-Part: RIG Proxy Metrics (#179)
Accenture/reactive-interaction-gateway: docs/metrics-details.md

## RIG Specific Metrics

### Proxy Metrics - `rig_proxy_requests_total` ###

With this metric the usage of RIGs API Proxy/Gateway can be monitored.

Following Labels are provided:
- **method** - The HTTP method called for this request Following possible values:
  - `GET`
  - `POST`
  - `PUT`
  - `PATCH`
  - `DELETE`
  - `HEAD`
  - `OPTIONS`
- **path** - The path used for this request.
- **target** - The target for this request. Grabbed from proxy parametrization. Following possible values:
  - `http`
  - `kafka`
  - `kinesis`
  - `N/A` - Not available, due to missing parametrization

> - **response_from** - Where the response is provided from. Grabbed from proxy parametrization. Following possible values:

Reviewer: imho "configuration", "config" or "settings" is a bit clearer/simpler than "parametrization"

  - `http`
  - `kafka`

>   - `N/A` - Not available, due to missing parametrization

Reviewer: ```suggestion
  - `N/A` - Not applicable, e.g., the request path is not configured.
```

- **status** - The *internal* status for the request. Attention: this status only tracks the internal rig process status. Once forwarded we track as "ok". We expect that called services are monitored on it's own. Following possible status codes:

>   - `ok` - If forwarded successfully

Reviewer: ```suggestion
  - `ok` - Forwarded successfully.
```

  - `ok` - if forwarded successfully

>   - `not_parameterized` - if the provided `method` and/or `path` isn't parameterized

Reviewer: as noted above, I think I'd go with not_found instead

Author: OK - I'll rename

>   - `bad_request` - If there are missing expected body parameters. Targets `kafka` and `kinesis` only.

Reviewer: ```suggestion
  - `bad_request` - Missing body parameters (`kafka` and `kinesis` only).
```

>   - `not_found` - If the provided `method` and/or `path` isn't found in the proxy parametrization

Reviewer: ```suggestion
  - `not_found` - No configuration found for the given `method` and/or `path`.
```

  - `bad_gateway` - Could not connect to the parameterized backend (e.g. offline)

>   - `method_not_allowed` - If HTTP method is used that is not provided in the list above

Reviewer: is there a reason to count this in addition to "not_found"?

Author: Hmm - I think yes. It could be the case, that a method IS parameterized, but NOTimplemented in the HTTPHandler

Author: as discussed - Removed. "Method_not_allowed" should be validated already on parametrization not on runtime.
I'll create an bug

>   - `unreachable` - Could not connect to the parameterized backend (e.g. offline)

Reviewer: ```suggestion
  - `unreachable` - The backend cannot be connected to (e.g., service is offline or domain cannot be resolved)
```

>   - `request_timeout` - If service is reachable, but times out.

Reviewer: ```suggestion
  - `request_timeout` - The service is reachable but fails to deliver a response in time.
```

>   - `gateway_timeout` - On timeout for async response. May only occur if endpoint parameter `response_from` is set to `kafka` or `kinesis`

Reviewer: ah okay so you treat connection timeouts towards the gateway as `bad_gateway` then, right? Not sure if we should use `gateway_timeout` then here, cause it has [a meaning](https://www.checkupdown.com/status/E504.html) in the http context.

Author: Hmm - Currently `:gateway_timeout` is also used in providing the http response to the client.

If we change, than we should change both, right?

Author: fixed as discussed offline

>   - `response_timeout` - On timeout for async response. May only occur if endpoint parameter `response_from` is set.

Reviewer: ```suggestion
  - `response_timeout` - The endpoint has its `response_from` parameter set and RIG times out while waiting for a response.
```
